candidate_base.go

```diff
@@ -496,103 +503,382 @@ func (c *candidateBase) Marshal() string {
 		c.Port(),
 		c.Type())
 
-	if c.tcpType != TCPTypeUnspecified {
-		val += fmt.Sprintf(" tcptype %s", c.tcpType.String())
-	}
-
 	if r := c.RelatedAddress(); r != nil && r.Address != "" && r.Port != 0 {
 		val = fmt.Sprintf("%s raddr %s rport %d",
 			val,
 			r.Address,
 			r.Port)
 	}
 
+	Extensions := c.Extensions()
+
+	if len(Extensions) > 0 {
+		val = fmt.Sprintf("%s %s", val, strings.Join(c.Extensions(), " "))
+	}
+
 	return val
 }
 
-// UnmarshalCandidate creates a Candidate from its string representation
+func (c *candidateBase) Extensions() []string {
+	extensions := make([]string, 0, len(c.extensions))
+	hasTCPTYPE := false
+
+	for i := 0; i < len(c.extensions); i += 2 {
+		key := c.extensions[i]
+
+		if i+1 >= len(c.extensions) {
+			// should never happen
```

Comment:
I am on phone, so don’t fully understand code yet. Can/should we return an error in bad situation?